File: litho-it/src/test/java/com/facebook/litho/MountStateRemountEventHandlerTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static com.facebook.litho.Column.create;
import static com.facebook.litho.MountState.getComponentClickListener;
import static com.facebook.litho.MountState.getComponentFocusChangeListener;
import static com.facebook.litho.MountState.getComponentLongClickListener;
import static com.facebook.litho.MountState.getComponentTouchListener;
import static com.facebook.litho.testing.helper.ComponentTestHelper.mountComponent;
import static org.assertj.core.api.Java6Assertions.assertThat;

import com.facebook.litho.config.TempComponentsConfigurations;
import com.facebook.litho.testing.inlinelayoutspec.InlineLayoutSpec;
import com.facebook.litho.testing.testrunner.LithoTestRunner;
import com.facebook.litho.widget.SimpleMountSpecTester;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(LithoTestRunner.class)
public class MountStateRemountEventHandlerTest {

  private ComponentContext mContext;

  @Before
  public void setup() {
    TempComponentsConfigurations.setShouldAddHostViewForRootComponent(true);
    mContext = new ComponentContext(getApplicationContext());
  }

  @Test
  public void testReuseClickListenerOnSameView() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .clickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentClickListener clickListener = getComponentClickListener(lithoView.getChildAt(0));
    assertThat(clickListener).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .clickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(clickListener == getComponentClickListener(lithoView.getChildAt(0))).isTrue();
  }

  @Test
  public void testReuseLongClickListenerOnSameView() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .longClickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentLongClickListener longClickListener =
        getComponentLongClickListener(lithoView.getChildAt(0));
    assertThat(longClickListener).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .longClickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(longClickListener == getComponentLongClickListener(lithoView.getChildAt(0)))
        .isTrue();
  }

  @Test
  public void testReuseFocusChangeListenerListenerOnSameView() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .focusChangeHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    ComponentFocusChangeListener focusChangeListener =
        getComponentFocusChangeListener(lithoView.getChildAt(0));
    assertThat(focusChangeListener).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .focusChangeHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(focusChangeListener == getComponentFocusChangeListener(lithoView.getChildAt(0)))
        .isTrue();
  }

  @Test
  public void testReuseTouchListenerOnSameView() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .touchHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentTouchListener touchListener = getComponentTouchListener(lithoView.getChildAt(0));
    assertThat(touchListener).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .touchHandler(c.newEventHandler(2))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(getComponentTouchListener(lithoView.getChildAt(0))).isEqualTo(touchListener);
  }

  @Test
  public void testUnsetClickHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .clickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    assertThat(getComponentClickListener(lithoView.getChildAt(0))).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    final ComponentClickListener listener = getComponentClickListener(lithoView);
    assertThat(listener).isNull();
  }

  @Test
  public void testUnsetLongClickHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .longClickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    assertThat(getComponentLongClickListener(lithoView.getChildAt(0))).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    final ComponentLongClickListener listener = getComponentLongClickListener(lithoView);
    assertThat(listener).isNull();
  }

  @Test
  public void testUnsetFocusChangeHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .focusChangeHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    assertThat(getComponentFocusChangeListener(lithoView.getChildAt(0))).isNotNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    final ComponentFocusChangeListener listener = getComponentFocusChangeListener(lithoView);
    assertThat(listener).isNull();
  }

  @Test
  public void testUnsetTouchHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .touchHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    final ComponentTouchListener listener = getComponentTouchListener(lithoView);
    assertThat(listener).isNull();
  }

  @Test
  public void testSetClickHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    assertThat(getComponentClickListener(lithoView)).isNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .clickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentClickListener listener = getComponentClickListener(lithoView.getChildAt(0));
    assertThat(listener).isNotNull();
    assertThat(listener.getEventHandler()).isNotNull();
  }

  @Test
  public void testSetLongClickHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    assertThat(getComponentLongClickListener(lithoView)).isNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .longClickHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentLongClickListener listener =
        getComponentLongClickListener(lithoView.getChildAt(0));
    assertThat(listener).isNotNull();
    assertThat(listener.getEventHandler()).isNotNull();
  }

  @Test
  public void testSetFocusChangeHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    assertThat(getComponentFocusChangeListener(lithoView)).isNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .focusChangeHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentFocusChangeListener listener =
        getComponentFocusChangeListener(lithoView.getChildAt(0));
    assertThat(listener).isNotNull();
    assertThat(listener.getEventHandler()).isNotNull();
  }

  @Test
  public void testSetTouchHandler() {
    final LithoView lithoView =
        mountComponent(
            mContext,
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(0);
    assertThat(getComponentTouchListener(lithoView)).isNull();

    lithoView
        .getComponentTree()
        .setRoot(
            new InlineLayoutSpec() {
              @Override
              protected Component onCreateLayout(ComponentContext c) {
                return create(c)
                    .touchHandler(c.newEventHandler(1))
                    .child(SimpleMountSpecTester.create(c))
                    .child(SimpleMountSpecTester.create(c))
                    .build();
              }
            });

    assertThat(lithoView.getChildCount()).isEqualTo(1);
    final ComponentTouchListener listener = getComponentTouchListener(lithoView.getChildAt(0));
    assertThat(listener).isNotNull();
    assertThat(listener.getEventHandler()).isNotNull();
  }

  @After
  public void restoreConfiguration() {
    TempComponentsConfigurations.restoreShouldAddHostViewForRootComponent();
  }
}

File: litho-it/src/test/java/com/facebook/litho/MountStateRemountEventHandlerTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.facebook.litho.config.TempComponentsConfigurations
import com.facebook.litho.testing.helper.ComponentTestHelper
import com.facebook.litho.testing.inlinelayoutspec.InlineLayoutSpec
import com.facebook.litho.testing.testrunner.LithoTestRunner
import com.facebook.litho.widget.SimpleMountSpecTester
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(LithoTestRunner::class)
class MountStateRemountEventHandlerTest {

  private lateinit var context: ComponentContext

  @Before
  fun setup() {
    TempComponentsConfigurations.setShouldAddHostViewForRootComponent(true)
    context = ComponentContext(ApplicationProvider.getApplicationContext<Context>())
  }

  @After
  fun restoreConfiguration() {
    TempComponentsConfigurations.restoreShouldAddHostViewForRootComponent()
  }

  @Test
  fun testReuseLongClickListenerOnSameView() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .longClickHandler(c.newEventHandler(1) as? EventHandler<LongClickEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(1)
    val longClickListener =
        LithoViewAttributesExtension.getComponentLongClickListener(lithoView.getChildAt(0))
    assertThat(longClickListener).isNotNull
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .longClickHandler(c.newEventHandler(1) as? EventHandler<LongClickEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(
            longClickListener ===
                LithoViewAttributesExtension.getComponentLongClickListener(lithoView.getChildAt(0)))
        .isTrue
  }

  @Test
  fun testReuseFocusChangeListenerListenerOnSameView() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .focusChangeHandler(c.newEventHandler(1) as? EventHandler<FocusChangedEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(1)
    val focusChangeListener =
        LithoViewAttributesExtension.getComponentFocusChangeListener(lithoView.getChildAt(0))
    assertThat(focusChangeListener).isNotNull
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .focusChangeHandler(c.newEventHandler(1) as? EventHandler<FocusChangedEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(
            focusChangeListener ===
                LithoViewAttributesExtension.getComponentFocusChangeListener(
                    lithoView.getChildAt(0)))
        .isTrue
  }

  @Test
  fun testReuseTouchListenerOnSameView() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .touchHandler(c.newEventHandler(1) as? EventHandler<TouchEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(1)
    val touchListener =
        LithoViewAttributesExtension.getComponentTouchListener(lithoView.getChildAt(0))
    assertThat(touchListener).isNotNull
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .touchHandler(c.newEventHandler(2) as? EventHandler<TouchEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(LithoViewAttributesExtension.getComponentTouchListener(lithoView.getChildAt(0)))
        .isEqualTo(touchListener)
  }

  @Test
  fun testUnsetClickHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .clickHandler(c.newEventHandler(1) as? EventHandler<ClickEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.getChildAt(0).isClickable).isTrue
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.isClickable).isFalse
  }

  @Test
  fun testUnsetLongClickHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .longClickHandler(c.newEventHandler(1) as? EventHandler<LongClickEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(1)
    assertThat(LithoViewAttributesExtension.getComponentLongClickListener(lithoView.getChildAt(0)))
        .isNotNull
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(0)
    val listener = LithoViewAttributesExtension.getComponentLongClickListener(lithoView)
    assertThat(listener).isNull()
  }

  @Test
  fun testUnsetFocusChangeHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .focusChangeHandler(c.newEventHandler(1) as? EventHandler<FocusChangedEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(1)
    assertThat(
            LithoViewAttributesExtension.getComponentFocusChangeListener(lithoView.getChildAt(0)))
        .isNotNull
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(0)
    val listener = LithoViewAttributesExtension.getComponentFocusChangeListener(lithoView)
    assertThat(listener).isNull()
  }

  @Test
  fun testUnsetTouchHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .touchHandler(c.newEventHandler(1) as? EventHandler<TouchEvent>)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(0)
    val listener = LithoViewAttributesExtension.getComponentTouchListener(lithoView)
    assertThat(listener).isNull()
  }

  @Test
  fun testSetClickHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(0)
    assertThat(lithoView.isClickable).isFalse
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .clickHandler(c.newEventHandler(1) as? EventHandler<ClickEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(1)
    assertThat(lithoView.getChildAt(0).isClickable).isTrue
  }

  @Test
  fun testSetLongClickHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(0)
    assertThat(LithoViewAttributesExtension.getComponentLongClickListener(lithoView)).isNull()
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .longClickHandler(c.newEventHandler(1) as? EventHandler<LongClickEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(1)
    val listener =
        LithoViewAttributesExtension.getComponentLongClickListener(lithoView.getChildAt(0))
    assertThat(listener).isNotNull
    assertThat(listener?.eventHandler).isNotNull
  }

  @Test
  fun testSetFocusChangeHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(0)
    assertThat(LithoViewAttributesExtension.getComponentFocusChangeListener(lithoView)).isNull()
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .focusChangeHandler(c.newEventHandler(1) as? EventHandler<FocusChangedEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(1)
    val listener =
        LithoViewAttributesExtension.getComponentFocusChangeListener(lithoView.getChildAt(0))
    assertThat(listener).isNotNull
    assertThat(listener?.eventHandler).isNotNull
  }

  @Test
  fun testSetTouchHandler() {
    // Needed because of (legacy) usage two different InlineLayoutSpec that the test wants to treat
    // as the same component type.
    val COMPONENT_IDENTITY = 12_345
    val lithoView =
        ComponentTestHelper.mountComponent(
            context,
            object : InlineLayoutSpec(COMPONENT_IDENTITY) {
              override fun onCreateLayout(c: ComponentContext): Component =
                  Column.create(c)
                      .child(SimpleMountSpecTester.create(c))
                      .child(SimpleMountSpecTester.create(c))
                      .build()
            })
    assertThat(lithoView.childCount).isEqualTo(0)
    assertThat(LithoViewAttributesExtension.getComponentTouchListener(lithoView)).isNull()
    lithoView.componentTree?.root =
        object : InlineLayoutSpec(COMPONENT_IDENTITY) {
          override fun onCreateLayout(c: ComponentContext): Component =
              Column.create(c)
                  .touchHandler(c.newEventHandler(1) as? EventHandler<TouchEvent>)
                  .child(SimpleMountSpecTester.create(c))
                  .child(SimpleMountSpecTester.create(c))
                  .build()
        }
    assertThat(lithoView.childCount).isEqualTo(1)
    val listener = LithoViewAttributesExtension.getComponentTouchListener(lithoView.getChildAt(0))
    assertThat(listener).isNotNull
    assertThat(listener?.eventHandler).isNotNull
  }
}
